### PR TITLE
Feature 11 pcl icp

### PR DESCRIPTION
**Pull request by Jebediah**
This PR is add a wrapper for PCL's ICP matcher, see #11 and #7. The covariance estimation is to come in a future PR, the focus for this one is the general matcher interface, should speed up PR's for other matchers

Summary:

- Add base class to derive specific matchers from
- Add ICP class that wraps PCL ICP
- Add PCL to the install scripts
- Add a few tests to try interface

### CODE REVIEW

**Comment by arunxdas** [comment]
PRs must also include api documentation

File: wave_matching/include/wave/matching/matcher.hpp [hunk 1]
```diff
@@ -0,0 +1,43 @@
+#ifndef __WAVE_MATCHING_MATCHER_HPP__
+#define __WAVE_MATCHING_MATCHER_HPP__
+
+#include <Eigen/Core>
+#include <Eigen/Geometry>
+
+namespace wave {
+namespace matching {
```
Comment (chutsu) [inline wave_matching/include/wave/matching/matcher.hpp]: I'm not sure we need the matching name space, I think we would like to avoid having additional namespaces if we can.

File: wave_matching/tests/icp_tests.cpp [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+class ICPTest : public testing::Test {
+  protected:
+    ICPTest() {}
+    virtual ~ICPTest() {}
+    virtual void SetUp() {
+        ref = boost::make_shared<pcl::PointCloud<pcl::PointXYZ> >();
+        target = boost::make_shared<pcl::PointCloud<pcl::PointXYZ> >();
+        pcl::io::loadPCDFile("../../wave_matching/tests/testscan.pcd", *ref);
+    }
+    void setparams(const float res, const Eigen::Affine3d perturb) {
```
Comment (chutsu) [inline wave_matching/tests/icp_tests.cpp]: Camel Case

File: wave_matching/tests/icp_tests.cpp [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+#include <gtest/gtest.h>
+#include <pcl/io/pcd_io.h>
+#include <Eigen/Dense>
+#include <iostream>
+
+#include "wave/matching/ICP.hpp"
+
+bool isApprox(const Eigen::Affine3d& first,
+              const Eigen::Affine3d& second,
+              double TOL) {
```
Comment (chutsu) [inline wave_matching/tests/icp_tests.cpp]: tolerance? why all upper case?

File: wave_matching/tests/icp_tests.cpp [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+  protected:
+    ICPTest() {}
+    virtual ~ICPTest() {}
+    virtual void SetUp() {
+        ref = boost::make_shared<pcl::PointCloud<pcl::PointXYZ> >();
+        target = boost::make_shared<pcl::PointCloud<pcl::PointXYZ> >();
+        pcl::io::loadPCDFile("../../wave_matching/tests/testscan.pcd", *ref);
+    }
+    void setparams(const float res, const Eigen::Affine3d perturb) {
+        matcher = new wave::matching::ICP_Matcher(res);
```
Comment (chutsu) [inline wave_matching/tests/icp_tests.cpp]: why `new`? where is delete?

File: wave_matching/tests/icp_tests.cpp [hunk 5]
```diff
@@ -0,0 +1,80 @@
+// Tests for wrapper
+
+#include <gtest/gtest.h>
+#include <pcl/io/pcd_io.h>
+#include <Eigen/Dense>
+#include <iostream>
+
+#include "wave/matching/ICP.hpp"
+
```
Comment (chutsu) [inline wave_matching/tests/icp_tests.cpp]: Consider putting this whole test file into the wave namespace

File: wave_matching/tests/icp_tests.cpp [hunk 6]
```diff
@@ -0,0 +1,80 @@
...
+    void setparams(const float res, const Eigen::Affine3d perturb) {
+        matcher = new wave::matching::ICP_Matcher(res);
+        pcl::transformPointCloud(*ref, *target, perturb);
+        matcher->setup(ref, target);
+    }
+    pcl::PointCloud<pcl::PointXYZ>::Ptr ref, target;
+    wave::matching::ICP_Matcher *matcher;
+};
+
+TEST(icp_tests, initialization) {
```
Comment (chutsu) [inline wave_matching/tests/icp_tests.cpp]: `icp_tests`? You mean `ICPTests`

File: wave_matching/src/ICP.cpp [hunk 7]
```diff
@@ -0,0 +1,57 @@
...
+        this->filter.setInputCloud(ref);
+        this->filter.filter(*(this->ref));
+    } else {
+        this->ref = ref;
+    }
+    this->icp.setInputSource(this->ref);
+}
+
+void ICP_Matcher::setTarget(const PCLPointCloud &target) {
+    if(resolution > 0) {
```
Comment (chutsu) [inline wave_matching/src/ICP.cpp]: `this->`

File: wave_matching/include/wave/matching/matcher.hpp [hunk 8]
```diff
@@ -0,0 +1,43 @@
...
+    virtual void setRef(const R& ref) = 0;
+    virtual void setTarget(const R& target) = 0;
+    void setup(const R& ref, const R& target) {
+        this->setRef(ref);
+        this->setTarget(target);
+    };
+    // intended to actually perform the matching, and
+    // then populate the private variables.
+    // should return true if successful
+    virtual bool match() {return 0;}
```
Comment (chutsu) [inline wave_matching/include/wave/matching/matcher.hpp]: Typically the method declaration should follow a logical order, with:

1. Constructors
2. Useful methods
3. Setters and Getters

you currently have 
1. Generic acccessors
2. Constructors
3. Generic accessors
4. Useful Methods

File: wave_matching/include/wave/matching/matcher.hpp [hunk 9]
```diff
@@ -0,0 +1,43 @@
...
+
+  protected:
+    // internal storage
```
Comment (chutsu) [inline wave_matching/include/wave/matching/matcher.hpp]: Superflous comment

File: wave_matching/tests/icp_tests.cpp [hunk 10]
```diff
@@ -0,0 +1,80 @@
...
+}
+
+// Small displacement using voxel downsampling
+TEST_F(ICPTest, smallDisplacement) {
+    Eigen::Affine3d perturb = Eigen::Affine3d::Identity();
+    perturb.translation() << 0.2, 0, 0;
+    setparams(0.05f, perturb);
+    EXPECT_TRUE(matcher->match());
+    Eigen::Affine3d result = matcher->get_result();
+    EXPECT_TRUE(isApprox(result, perturb, 0.1));
```
Comment (chutsu) [inline wave_matching/tests/icp_tests.cpp]: everything in your tests are a little muddle up and hard to read, I personally would have done it this way:

```
Eigen::Affine3d perturb;
Eigen::Affine3d result 

// setup
perturb = Eigen::Affine3d::Identity();
perturb.translation() << 0.2, 0, 0;
setparams(0.05f, perturb);

// test and assert
result = matcher->get_result();
EXPECT_TRUE(matcher->match());
EXPECT_TRUE(isApprox(result, perturb, 0.1));
```

File: wave_matching/tests/icp_tests.cpp [hunk 11]
```diff
@@ -0,0 +1,80 @@
...
+
+bool isApprox(const Eigen::Affine3d& first,
+              const Eigen::Affine3d& second,
+              double TOL) {
+    for (uint8_t iii = 0; iii < 4; ++iii) {
+        for(uint8_t jjj = 0; jjj < 4; ++jjj) {
+            TOL -= std::abs(first(iii,jjj) - second(iii,jjj));
+        }
+    }
+    if (TOL > 0) {
```
Comment (chutsu) [inline wave_matching/tests/icp_tests.cpp]: Floating point comparisons, this may never be true. http://floating-point-gui.de/errors/comparison/

File: wave_matching/tests/icp_tests.cpp [hunk 12]
```diff
@@ -0,0 +1,80 @@
...
+#include <pcl/io/pcd_io.h>
+#include <Eigen/Dense>
+#include <iostream>
+
+#include "wave/matching/ICP.hpp"
+
+bool isApprox(const Eigen::Affine3d& first,
+              const Eigen::Affine3d& second,
+              double TOL) {
+    for (uint8_t iii = 0; iii < 4; ++iii) {
```
Comment (chutsu) [inline wave_matching/tests/icp_tests.cpp]: Why `uint8_t` why not normal int?

File: wave_matching/tests/icp_tests.cpp [hunk 13]
```diff
@@ -0,0 +1,80 @@
...
+    return false;
+}
+
+// Fixture to load same pointcloud all the time
+class ICPTest : public testing::Test {
+  protected:
+    ICPTest() {}
+    virtual ~ICPTest() {}
+    virtual void SetUp() {
+        ref = boost::make_shared<pcl::PointCloud<pcl::PointXYZ> >();
```
Comment (chutsu) [inline wave_matching/tests/icp_tests.cpp]: `this->`

File: wave_matching/tests/icp_tests.cpp [hunk 14]
```diff
@@ -0,0 +1,80 @@
...
+        target = boost::make_shared<pcl::PointCloud<pcl::PointXYZ> >();
```
Comment (chutsu) [inline wave_matching/tests/icp_tests.cpp]: `this->`

File: wave_matching/tests/icp_tests.cpp [hunk 15]
```diff
@@ -0,0 +1,80 @@
+// Tests for wrapper
+
+#include <gtest/gtest.h>
+#include <pcl/io/pcd_io.h>
+#include <Eigen/Dense>
+#include <iostream>
```
Comment (chutsu) [inline wave_matching/tests/icp_tests.cpp]: header file ordering is:

1. C 
2. C++ 
3. Third Party 
4. Internal

File: wave_matching/include/wave/matching/matcher.hpp [hunk 16]
```diff
@@ -0,0 +1,43 @@
...
+
+namespace wave {
+namespace matching {
+
+// sensor data type
+template <typename R>
+class Matcher {
+  public:
+    // generic accessors
+    const Eigen::Affine3d& get_result() {return result;};
```
Comment (chutsu) [inline wave_matching/include/wave/matching/matcher.hpp]: - `this->result`
- `getResult()`

File: wave_matching/include/wave/matching/matcher.hpp [hunk 17]
```diff
@@ -0,0 +1,43 @@
...
+    const Eigen::MatrixXd& get_info() {return information;};
```
Comment (chutsu) [inline wave_matching/include/wave/matching/matcher.hpp]: - `this->information`
- `getInfo()`

File: wave_matching/include/wave/matching/matcher.hpp [hunk 18]
```diff
@@ -0,0 +1,43 @@
...
+    const float& getRes() {return resolution;};
```
Comment (chutsu) [inline wave_matching/include/wave/matching/matcher.hpp]: - Why a reference `&` to the float? just return float?
- `this->resolution`

File: wave_matching/include/wave/matching/matcher.hpp [hunk 19]
```diff
@@ -0,0 +1,43 @@
...
+#define __WAVE_MATCHING_MATCHER_HPP__
+
+#include <Eigen/Core>
+#include <Eigen/Geometry>
+
+namespace wave {
+namespace matching {
+
+// sensor data type
+template <typename R>
```
Comment (chutsu) [inline wave_matching/include/wave/matching/matcher.hpp]: Would have used `T`instead of `R` for typename, unless you have a good reason to choose `R`?

File: wave_matching/tests/icp_tests.cpp (hunk 2 above)
Comment (michaelhsmart) [inline wave_matching/tests/icp_tests.cpp]: camelCase

Comment (chutsu) [inline wave_matching/tests/icp_tests.cpp]: actually both camelCase and CamelCase are both camel case.

File: wave_matching/tests/icp_tests.cpp (hunk 11 above)
Comment (Jebediah) [inline wave_matching/tests/icp_tests.cpp]: But it's not checking if it is equal to 0, just greater than. Shouldn't that work even with the floating point error?

File: wave_matching/include/wave/matching/matcher.hpp (hunk 18 above)
Comment (Jebediah) [inline wave_matching/include/wave/matching/matcher.hpp]: why not?

File: wave_matching/tests/icp_tests.cpp (hunk 5 above)
Comment (Jebediah) [inline wave_matching/tests/icp_tests.cpp]: why?

File: wave_matching/tests/icp_tests.cpp (hunk 2 above)
Comment (michaelhsmart) [inline wave_matching/tests/icp_tests.cpp]: Full pedant: lowerCamelCase vs UpperCamelCase

:)

File: wave_matching/tests/icp_tests.cpp (hunk 12 above)
Comment (Jebediah) [inline wave_matching/tests/icp_tests.cpp]: It's only going to 3, seemed like a waste of an int

Comment (chutsu) [inline wave_matching/tests/icp_tests.cpp]: pre-mature optimization is the root of all evil.

File: wave_matching/tests/icp_tests.cpp (hunk 11 above)
Comment (chutsu) [inline wave_matching/tests/icp_tests.cpp]: ah touche, was too quick on this one.

File: wave_matching/tests/icp_tests.cpp [hunk 20]
```diff
@@ -0,0 +1,80 @@
...
+
+#include "wave/matching/ICP.hpp"
+
+bool isApprox(const Eigen::Affine3d& first,
+              const Eigen::Affine3d& second,
+              double TOL) {
+    for (uint8_t iii = 0; iii < 4; ++iii) {
+        for(uint8_t jjj = 0; jjj < 4; ++jjj) {
+            TOL -= std::abs(first(iii,jjj) - second(iii,jjj));
```
Comment (michaelhsmart) [inline wave_matching/tests/icp_tests.cpp]: -= on a pass by value argument?

Does that not trigger a compiler warning?

Comment (Jebediah) [inline wave_matching/tests/icp_tests.cpp]: Should it? It is a copy so it's not going to hurt the original.

Comment (stevenwaslander) [inline wave_matching/tests/icp_tests.cpp]: Can we not use iii and jjj in for loops?

Comment (michaelhsmart) [inline wave_matching/tests/icp_tests.cpp]: I would have assumed it would. I can't imagine intended cases of that are as common as bug cases.

Comment (michaelhsmart) [inline wave_matching/tests/icp_tests.cpp]: This was an honest question, btw. Not a review.

File: wave_matching/tests/icp_tests.cpp (hunk 2 above)
Comment (Jebediah) [inline wave_matching/tests/icp_tests.cpp]: please

File: wave_matching/include/wave/matching/matcher.hpp (hunk 18 above)
Comment (chutsu) [inline wave_matching/include/wave/matching/matcher.hpp]: because:
1. float is trivially small, pass by reference is good when you're passing large objects
2. you're returning a reference of `resolution` rather than value, but I'm not actually certain if `resolution` changes, it would also change the reference, e.g.

```
// suppose I do
float r;
r = Matcher.getRes();
// Matcher.resolution changes
r; // did r change as well? I'm not certain r will change as well, so why not just return a value instead of a reference
```

File: wave_matching/tests/icp_tests.cpp (hunk 2 above)
Comment (stevenwaslander) [inline wave_matching/tests/icp_tests.cpp]: So, what is going on here? Can someone just suggest the correction.  setParams, right?

File: wave_matching/include/wave/matching/matcher.hpp (hunk 18 above)
Comment (stevenwaslander) [inline wave_matching/include/wave/matching/matcher.hpp]: I would suggest to keep things simple.  Don't unecessarily pass by reference such a tiny thing.  Another optimization that is not warranted.

File: wave_matching/tests/icp_tests.cpp (hunk 5 above)
Comment (chutsu) [inline wave_matching/tests/icp_tests.cpp]: To save you from having to prepend `wave::` infront of your wave functions / classes. Plus all tests will be using the namespace `wave` so to be consistent I would advise you do the same even though you only use it once here.

File: wave_matching/include/wave/matching/matcher.hpp [hunk 21]
```diff
@@ -0,0 +1,43 @@
...
+class Matcher {
+  public:
```
Comment (stevenwaslander) [inline wave_matching/include/wave/matching/matcher.hpp]: Should be part of the checklist, used clang? check.  Was there a checklist?

File: wave_matching/tests/icp_tests.cpp (hunk 20 above)
Comment (chutsu) [inline wave_matching/tests/icp_tests.cpp]: w.r.t `TOL -= ..` its fine in this case.

File: scripts/install/pcl_install.bash [hunk 22]
```diff
@@ -0,0 +1,19 @@
...
+        libvtk6-dev \
+        libeigen3-dev \
+        libusb-1.0-0-dev
+}
+
+echo "Installing PCL ..."
+
+install_dependencies
+
+sudo add-apt-repository -y ppa:v-launchpad-jochen-sprickerhof-de/pcl
```
Comment (stevenwaslander) [inline scripts/install/pcl_install.bash]: Do we have to use a PPA version? should we be compiling from source instead or using the one in ROS Kinetic?

File: .gitignore [hunk 23]
```diff
@@ -21,3 +21,7 @@ build
 # matlab / octave
 *.mat
 octave-workspace
+
+# clion
+.idea/
+cmake-build-*/
```
Comment (stevenwaslander) [inline .gitignore]: Are these changes required for this PR? Are they related? I expect we want a universal .gitignore that includes them, but we want to avoid changing them to whatever people want to use locally on random PRs. Don't we have a standard for .gitignore?

File: wave_matching/tests/icp_tests.cpp (hunk 12 above)
Comment (chutsu) [inline wave_matching/tests/icp_tests.cpp]: 1 last thing, admittedly stupid detail its up to you, but consider using `i++` instead of `++i`. While both are perfectly valid I rarely (if ever) see `++i`. 

Once you're outside the norm, subtle off by one errors creep in. For example these two loops do different things.

```
int i = 0;
int j = i;
while(j < 5) {
    printf("%d", i);
    j = ++i;
}
// prints 1 2 3 4

int i = 0;
int j = i;
while(j < 5) {
    printf("%d", i);
    j = i++;
}
// prints 0 1 2 3 4 5 
```

http://ideone.com/U3Wz2K

File: scripts/install/pcl_install.bash (hunk 22 above)
Comment (Jebediah) [inline scripts/install/pcl_install.bash]: It is from PCL's website: http://pointclouds.org/downloads/linux.html

Comment (Jebediah) [inline scripts/install/pcl_install.bash]: I don't think we should use the one in ROS Kinetic since libwave is supposed to be ROS-free. We could build from source, but it makes travis slower.

File: .gitignore (hunk 23 above)
Comment (Jebediah) [inline .gitignore]: These are for ignoring some garbage created by the IDE I am using (clion).

Comment (chutsu) [inline .gitignore]: quite a few lab members use clion so that addition to gitignore is fine. We currently don't have a universal `.gitignore`, I personally don't see the problem adding custom gitignore rules as they come (since they are pretty rare).

File: scripts/install/pcl_install.bash (hunk 22 above)
Comment (stevenwaslander) [inline scripts/install/pcl_install.bash]: OK.

File: .gitignore (hunk 23 above)
Comment (stevenwaslander) [inline .gitignore]: OK.

File: wave_matching/CMakeLists.txt [hunk 24]
```diff
@@ -0,0 +1,40 @@
+CMAKE_MINIMUM_REQUIRED(VERSION 2.8.3)
+PROJECT(wave_matching)
+
+IF (CMAKE_COMPILER_IS_GNUCXX)
+    ADD_DEFINITIONS(-std=gnu++11)
+ENDIF()
```
Comment (chutsu) [inline wave_matching/CMakeLists.txt]: this is defined in the master branch in root `CMakeFile.txt` so this is not needed.

Comment (chutsu) [inline wave_matching/CMakeLists.txt]: this is defined in the master branch in global `CMakeFile.txt` so this is not needed.

File: wave_matching/CMakeLists.txt [hunk 25]
```diff
@@ -0,0 +1,40 @@
...
+        /usr/lib
+        /usr/local/lib)
+
+set(DEPS
+    ${PCL_COMMON_LIBRARIES}
+    ${PCL_REGISTRATION_LIBRARIES}
+    ${PCL_POINT_TYPES_LIBRARIES}
+    ${PCL_IO_LIBRARIES})
+
+add_library(wave_matching STATIC src/ICP.cpp)
```
Comment (chutsu) [inline wave_matching/CMakeLists.txt]: I believe in our style guide we say source files should all be lower case separated by an underscore.

File: wave_matching/tests/icp_tests.cpp [hunk 26]
```diff
@@ -23,30 +25,34 @@ bool isApprox(const Eigen::Affine3d &first,
...
     virtual void SetUp() {
         this->ref = boost::make_shared<pcl::PointCloud<pcl::PointXYZ> >();
         this->target = boost::make_shared<pcl::PointCloud<pcl::PointXYZ> >();
         pcl::io::loadPCDFile("../../wave_matching/tests/testscan.pcd",
                              *(this->ref));
     }
+
     void setParams(const float res, const Eigen::Affine3d perturb) {
-        this->matcher = new wave::ICP_Matcher(res);
+        this->matcher = new ICP_Matcher(res);
```
Comment (chutsu) [inline wave_matching/tests/icp_tests.cpp]: No underscore

File: wave_matching/tests/icp_tests.cpp [hunk 27]
```diff
@@ -23,30 +25,34 @@ bool isApprox(const Eigen::Affine3d &first,
 
 // Fixture to load same pointcloud all the time
 class ICPTest : public testing::Test {
-protected:
+ protected:
     ICPTest() {}
+
     virtual ~ICPTest() {
         if (matcher) {
```
Comment (chutsu) [inline wave_matching/tests/icp_tests.cpp]: `this->`

File: wave_matching/tests/icp_tests.cpp [hunk 28]
```diff
@@ -23,30 +25,34 @@ bool isApprox(const Eigen::Affine3d &first,
...
             delete matcher;
```
Comment (chutsu) [inline wave_matching/tests/icp_tests.cpp]: `this->`

File: wave_matching/include/wave/matching/ICP.hpp [hunk 29]
```diff
@@ -0,0 +1,30 @@
...
+#include <pcl/point_types.h>
+#include <pcl/point_cloud.h>
+#include <pcl/registration/icp.h>
+#include <pcl/filters/voxel_grid.h>
+#include "wave/matching/matcher.hpp"
+
+namespace wave {
+
+typedef pcl::PointCloud<pcl::PointXYZ>::Ptr PCLPointCloud;
+class ICP_Matcher : public wave::Matcher<PCLPointCloud> {
```
Comment (chutsu) [inline wave_matching/include/wave/matching/ICP.hpp]: under `wave` namespace already don't need wave here

File: wave_matching/include/wave/matching/matcher.hpp [hunk 30]
```diff
@@ -0,0 +1,51 @@
...
+
+#include <Eigen/Core>
+#include <Eigen/Geometry>
+
+namespace wave {
+
+// sensor data type
+template <typename T>
+class Matcher {
+ public:
```
Comment (chutsu) [inline wave_matching/include/wave/matching/matcher.hpp]: you didn't run clang-format on this file, else `public` would not be indented

File: wave_matching/tests/icp_tests.cpp [hunk 31]
```diff
@@ -0,0 +1,111 @@
...
+
+// Zero displacement without downsampling
+TEST_F(ICPTest, fullResNullMatch) {
+    Eigen::Affine3d perturb;
+    Eigen::Affine3d result;
+    bool match_success = false;
+    // Setup
+    perturb = Eigen::Affine3d::Identity();
+    perturb.translation() << 0, 0, 0;
+    setParams(-1, perturb);
```
Comment (chutsu) [inline wave_matching/tests/icp_tests.cpp]: `this->`

File: wave_matching/tests/icp_tests.cpp [hunk 32]
```diff
@@ -0,0 +1,111 @@
...
+    EXPECT_TRUE(match_success);
+    EXPECT_TRUE(isApprox(result, perturb, 0.1));
+}
+
+}  // end of namespace wave
+
+int main(int argc, char *argv[]) {
+    testing::InitGoogleTest(&argc, argv);
+    return RUN_ALL_TESTS();
+}
```
Comment (chutsu) [inline wave_matching/tests/icp_tests.cpp]: At current I have a general wave_utils_tests and wave_optimization_tests test runner for each module of libwave, maybe you should do the same? For example this is how I link up all my test cases into a single test runner in the utils module:

```
# UNIT TESTS
SET(CMAKE_RUNTIME_OUTPUT_DIRECTORY ${PROJECT_BINARY_DIR}/tests)
FILE(COPY tests/data DESTINATION ${PROJECT_BINARY_DIR}/tests)

ADD_EXECUTABLE(
    # TARGET
    wave_utils_tests
    # TESTS
    tests/utils/config_test.cpp
    tests/utils/data_test.cpp
    tests/utils/filesystem_test.cpp
    tests/utils/math_test.cpp
    tests/utils/time_test.cpp
    # TEST RUNNER
    tests/test_runner.cpp
)
TARGET_LINK_LIBRARIES(wave_utils_tests ${TEST_DEPS})
```
Open to suggestions if you have alternatives.

File: wave_matching/CMakeLists.txt (hunk 24 above)
Comment (Jebediah) [inline wave_matching/CMakeLists.txt]: ok

File: wave_matching/include/wave/matching/matcher.hpp (hunk 30 above)
Comment (Jebediah) [inline wave_matching/include/wave/matching/matcher.hpp]: yes I "fixed" that part manually. Google style guide says to have a 1-space indent on labels inside classes.

File: wave_matching/tests/icp_tests.cpp (hunk 32 above)
Comment (Jebediah) [inline wave_matching/tests/icp_tests.cpp]: I like this. It will scale well with more matchers.

File: wave_matching/include/wave/matching/matcher.hpp (hunk 30 above)
Comment (chutsu) [inline wave_matching/include/wave/matching/matcher.hpp]: I think we might want to have an exception to that rule. I'm in favor of no indent, unless you have problems with it? 

@michaelhsmart? thoughts?

Comment (chutsu) [inline wave_matching/include/wave/matching/matcher.hpp]: ok mike says nay.

Can you change the `.clang-format` rule from `AccessModifierOffset: -4` to `AccessModifierOffset: -3`

Comment (Jebediah) [inline wave_matching/include/wave/matching/matcher.hpp]: no problem, I couldn't care less about how many spaces a label gets. Exception just needs to be documented.

Comment (chutsu) [inline wave_matching/include/wave/matching/matcher.hpp]: We may revisit this later :D

Comment (michaelhsmart) [inline wave_matching/include/wave/matching/matcher.hpp]: commenting here so it's tracked:

We've used 1 for autonomoose to keep cpplint from spewing piles except where we truly do deviate.

Depends solely on your desired level of cpplint adherence.

**Comment by chutsu** [comment]
When merging please edit the commit bullet points to only KEY ONES and select "Squash and Merge"
